### PR TITLE
[XML] Preferences page for validation XML

### LINKED ISSUE #634: [XML] Preferences page for validation XML
To fix some XML validation errors described in https://github.com/eclipse/wildwebdeveloper/issues/557, WWD should provide the capability to customize XML LemMinx validation. To do that WWD should provide the UI preferences to customize those settings.

### PR DESCRIPTION
**Pull request by angelozerr**
[XML] Preferences page for validation XML

Fixes #634

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
I need to clean my code but validation preferences page is implemented with all lemmings settings for validation:

image

The default settings for those validation fix 2 issues coming from https://github.com/eclipse/wildwebdeveloper/issues/557

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/messages.properties [hunk 1]
```diff
@@ -10,14 +10,24 @@
...
 
+# XML Catalog preference page
 XMLCatalogPreferencesPage_Entries=XML Catalog Entries:
 XMLCatalogPreferencesPage_Edit=Open in &Editor...
 XMLCatalogPreferencesPage_OpenInEditorTitle=Open XML Catalog in editor
 XMLCatalogPreferencesPage_OpenInEditorMessage=Do you want to Apply the changes and open the selected XML Catalog in editor?
 XMLCatalogPreferencesPage_OpenInEditorApplyAndEdit=Apply and Edit
 XMLCatalogPreferencesPage_OpenInEditorNo=No
 
+# XML Validation preference page
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/messages.properties:24]: Can you please add mnemonics here?

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * Contributors:
+ *  Angelo ZERR (Red Hat Inc.) - initial implementation
+ *******************************************************************************/
+package org.eclipse.wildwebdeveloper.xml.internal.ui.preferences;
+
+import org.eclipse.jface.util.PropertyChangeEvent;
+
+public class XMLPreferenceConstants {
+
+	public static final String XML_PREFERENCES_CATAGLOGS = "wildwebdeveloper.xml.catalogs";
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java:19]: Why keeping the `wildwebdeveloper` prefix here? the preference story should already be bound to the bundle.

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Angelo ZERR (Red Hat Inc.) - initial implementation
+ *******************************************************************************/
+package org.eclipse.wildwebdeveloper.xml.internal.ui.preferences;
+
+import org.eclipse.jface.util.PropertyChangeEvent;
+
+public class XMLPreferenceConstants {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java:17]: Please comment that there is a convention between Eclipse preference id and the LemMinX option (eg `foo.bar.lol` for `{ "foo": { "bar": { "lol": value }}}`

**Comment by mickaelistria** [comment]
Thanks a lot for this great PR. Please see comments inline. Most important one I'd like you to discuss is the one about properties name; other ones -mnemonics, comments...- I can take care of them.

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java:19]: Today catalogs is stored with this preference, but I can change that if you wish.

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java:28]: fixed, I add a link to vscode-xml which references all LemMinX settings.

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLPreferenceConstants.java]: fixed.

File: org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/messages.properties (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/messages.properties:24]: fixed

**Comment by angelozerr** [comment]
> Thanks a lot for this great PR.

You are welcome! It will fix a lot of problems reported by user, I will explain that in each issues when the PR will be merged.

Once this PR will be merged, I will add other preferences like codelens, format in order to promote the LemMinx features.

**Comment by mickaelistria** [comment]
I'm afraid that trying to map Wild Web Developer preference ids with LemMinX preference is safe: if lemminx preferences change, prefs in Wild Web Developer are broken and we have to implement a form of migration; and the other way round, if we want to add or modify a preference id in Wild Web Developer, we suddenly break the binding with LemMinX.
I'd rather see LemMinXPreference be a full type with fields like `preferenceId` & `lemminxOptionPath` and even the `register` part of this type. I can take care of implementing that if you're busy with other things.

**Comment by angelozerr** [comment]
@mickaelistria I did this PR in my free time, because I would like to fix all validation issues from WWD. 

If you could continue to improve my PR, it should be really nice. After that I could create other PRs quickly for codelens,format etc by following your idea that you suggested. Thanks!

**Comment by mickaelistria** [comment]
Too bad, I just pushed the previous commit + a refactoring. Can you please rebase the work you did in last commit on top of master?

**Comment by angelozerr** [comment]
Sorry @mickaelistria  -( Please ignore my PR. But XMLPreferenceInitializer has 2 methods which is not used.

**Comment by mickaelistria** [comment]
I've incorporated your change. I'll ping you later today or tomorrow when a build is available for you to verify I didn't do something too wrong. Not that we're missing tests covering XML options and catalogs, so it's very regression-prone. I'll try to improve that by the way.

**Pull request closed without merging**